RFC 2047 fixup

niklasha opened the pull request:
RFC2047 chapter 5 talks about encoding words, but the existing encoding both coalesced words as well as tried to just encode substrings of words, causing at least two mail user agents (Thunderbird and Apple Mail) to misrepresent subjects.  Instead take care to only encode single words.

Esp. point (1) says:

Ordinary ASCII text and 'encoded-word's may appear together in the
    same header field.  However, an 'encoded-word' that appears in a
    header field defined as '*text' MUST be separated from any adjacent
    'encoded-word' or 'text' by 'linear-white-space'.

This was not what the former code was doing, e.g. the word "Administratör" would separate into three: "administrat" "ö" and "r". and even if the encoded representation was without space in between, at least two well known MUAs (Thunderbird and Applemail") would insert whitespace in between.

paolobarbolini commented:
Thanks for the PR. I don't get why you removed the encoding buffering though. The only problem we had with the RFC is the fact that we didn't leave white space between the ascii text and the encoded part right?

niklasha commented:
> Thanks for the PR. I don't get why you removed the `encode_buf` part though. The only problem we had with the RFC is the fact that we didn't leave white space between the ascii text and the encoded part right?

No rather the opposite. It is renderers (like MUAs), that renders whitespace between parts that should not have it.  I discovered it when trying to mail with a Subject that contained the word "Ambassadör", which got encoded sth like "Ambassad=?utf-8?b?w7Y=?=r" (i.e. no whitespace in the encoding), which is then by both Apple Mail and Thunderbird , decoded and represented as "Ambassad ör".  So the critical part is that when special characters are in a word, **all** of the word need to be encoded.  It is possible that you are right that it is alwasy safe to include several words in the encoding, but then my concern was that the whitespace contained there would be "non breaking", so I just removed that running encoding as it also was added complexity.  I have no problem with you keeping that if you want/need to, then probably some unit tests need to be adapted again.  I did not see any problem by that code, it was just when I fixed this, I wanted to do it simple :-)

paolobarbolini commented:
I think the real problem might be here

https://github.com/lettre/lettre/blob/d6128a146e72d1cb3c5a66d84dbe924fd7fb3b87/src/message/header/mod.rs#L394-L408

niklasha commented:
Yes, that is definitely the problematic part as it offsets into words, hence splitting them leaving some parts unencoded and the middle part encoded, which is a nono according to the specs which tells us encoded parts must be separated from other parts by whitespace.  My take on this was that since we already have word-splitted input, just encode the words that need encoding.  But I trust you see the problem now, and can make a solution you like.

paolobarbolini commented:
I'm testing out a solution locally

paolobarbolini commented:
I've created #811. It still has some problems, as the email_encoding crate seems to be missing some line folding opportunities :thinking:. I'll look into it deeper in the following days.

paolobarbolini commented:
Done in #811
